### PR TITLE
Windows icons support

### LINKED ISSUE #16: Get file icon on windows
Find a way to read file icon on Windows

### PR DESCRIPTION
**Pull request by SkaceKamen**
Uses [win-icon-extractor](https://github.com/SkaceKamen/win-icon-extractor) package (created specially for this). This should resolve #16.

**Test build (64bit):**
https://dl.dropboxusercontent.com/u/85747523/Cerebro%20Setup%200.2.6.exe

**Features:**
- Works for all files and folders (not only binaries)

**Limitations:**
 - icons are only 32x32 (this is what windows libraries supports right now)
 - only works on windows (no problem here)

**Notes:**
 - since this module uses ffi library, which uses native node modules, `yarn run rebuild` in app folder is required to correctly compile native modules

2017-03-01 21_53_28-nvidia geforce overlay

### CODE REVIEW

File: app/lib/getFileIcon/index.js [hunk 1]
```diff
@@ -1,3 +1,6 @@
 const empty = () => Promise.reject()
-const getFileIcon = process.platform === 'darwin' ? require('./mac') : empty
+const getFileIcon = process.platform === 'darwin' ? require('./mac') : (
```
Comment (KELiON) [inline app/lib/getFileIcon/index.js]: I'd change it to `if/else if/else`. Ternary operator looks good only with two branches:)

File: app/lib/getFileIcon/windows.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+/**
+ * Get system icon for file
+ *
+ * @param  {String} path File path
+ * @param  {Number} options.width
+ * @param  {[type]} options.height
+ * @return {Promise<String>} Promise resolves base64-encoded source of icon
+ */
+export default function getFileIcon(path, { width = 128, height = 128 } = {}) {
```
Comment (KELiON) [inline app/lib/getFileIcon/windows.js]: You are not using second argument (object with width/height), so you can remove it (+ update doc of this method)

File: app/lib/getFileIcon/windows.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  return new Promise((resolve, reject) => {
+    var icon = null;
+    try {
+      icon = extractIcon(path)
```
Comment (KELiON) [inline app/lib/getFileIcon/windows.js]: Here you can just do `resolve(extractIcon(path))`, and remove `icon` variable

File: app/lib/getFileIcon/windows.js (hunk 2 above)
Comment (SkaceKamen) [inline app/lib/getFileIcon/windows.js]: Will do.

File: app/lib/getFileIcon/windows.js (hunk 3 above)
Comment (SkaceKamen) [inline app/lib/getFileIcon/windows.js]: Yes, the code was different originally, but it isn't required now.

File: app/lib/getFileIcon/index.js (hunk 1 above)
Comment (SkaceKamen) [inline app/lib/getFileIcon/index.js]: You're right, I just patched it to work and forget it's there :(

**Comment by SkaceKamen** [comment]
The size including the dependencies is 34 MB, the biggest library being `node-ffi`, which is same library `nodobjc` uses to extract the icons on mac.
I may be able to get rid of about 10 MB by removing dependency on image manipulation library, which I "only" use to convert bmp files to png and write the code myself / search for something smaller.

**Comment by KELiON** [comment]
@SkaceKamen also after all fixes would be cool if you can share final build (you can do it with `yarn run package`) here, so someone else could test it on their machines

File: app/lib/getFileIcon/index.js [hunk 4]
```diff
@@ -1,3 +1,19 @@
...
+}
+
+if (process.platform === 'win32') {
+  getFileIcon = require('./windows')
+}
+
 export default getFileIcon
+
+/* eslint-enable global-require */
+/* eslint-disable import/no-mutable-exports */
```
Comment (SkaceKamen) [inline app/lib/getFileIcon/index.js]: Don't know it this is the best way, but it sure was the easiest.

**Comment by SkaceKamen** [comment]
Ok, here's the installer for x64:
https://dl.dropboxusercontent.com/u/85747523/Cerebro%20Setup%200.2.6.exe
(is there any easy way to build installer for both x64 and 32bit ?

**Comment by AlexAlabyev** [comment]
@SkaceKamen Just tested your build. It works fine on Windows 10

**Comment by KELiON** [comment]
@AlexAlabyev would you mind to test windows build above?

**Comment by AlexAlabyev** [comment]
@KELiON sure. I've just tested it, and it works fine, icons are being displayed, and there's no errors in the console.

**Comment by KELiON** [comment]
@SkaceKamen is it correct that we don't have icons in Preview pane because we can get only 32x32 px? Is there a way to get bigger icon for Preview pane?

**Comment by SkaceKamen** [comment]
The current method I'm using always use 32x32, but it may be possible to extract bigger icons with different approach, I'lI investigate.

**Comment by KELiON** [comment]
@SkaceKamen actually, I don't think it is necessary. I have doubts about relevancy of application Preview pane – we show there big icon and file size information. But it looks unneeded – who cares about the size of the application when launches it? 

So I'd just use 32x32px, and I think we can remove preview in application plugin

**Comment by AlexAlabyev** [comment]
@KELiON That's a good point. There's no need in that preview when all you're trying to do is to open an app. It's better to leave it as `width:100%` and use 32x32 as an app icon on the left

**Comment by SkaceKamen** [comment]
It's probably fine leaving 32x32, but bigger icons would help on high DPI displays.

**Comment by SkaceKamen** [comment]
Anyway, I researched how to load bigger icons and it would involve a lot more work than current solution, so maybe in future, but right now it's too much work.

**Comment by AlexAlabyev** [comment]
@SkaceKamen 32x32 looks fine on 1080p, I don't think It would look blurry on 4K either, but if it does, it would be easier to just put a 10% - 20% sharpen filter on top of that 32x32 icon for the high DPI screens. It would have pretty much the same result visually

**Comment by SkaceKamen** [comment]
I updated the library, it now supports unicode paths (japanese, russian etc) and is more stable (there was small posibility of failing or proper path).

**Comment by SkaceKamen** [comment]
The travis CI failed because:
```
npm ERR! notsup Unsupported platform for win-icon-extractor@1.0.4: wanted {"os":"win32","arch":"any"} (current: {"os":"linux","arch":"x64"})
```
so I moved win-icon-extractor to optional dependencies, which won't stop installation when it's skipped.

**Comment by Stanzilla** [comment]
Very nice, can't wait for this to land!

**Comment by KELiON** [comment]
This is awesome! Thanks for doing it, @SkaceKamen!

**Comment by Stanzilla** [comment]
@KELiON any ETA when you will do the next release?

**Comment by KELiON** [comment]
@Stanzilla as soon as we finish #230 and test it

**Comment by KELiON** [comment]
@SkaceKamen looks like electron recently added API for icons: https://github.com/electron/electron/pull/7851. If it works like I think we can use it instead of adding module in dependencies.

**Comment by SkaceKamen** [comment]
That would be better solution, is it already available?

**Comment by KELiON** [comment]
@SkaceKamen I just updated electron version so it is available. I also tried it on MacOS and it doesn't work as I expected: it always show icon, associated with file type. So, instead of showing app icon it just shows generic app icon. 

But maybe on windows it works other way, so we can try it.

**Comment by SkaceKamen** [comment]
I looked at the code and it should work as expected on windows, I'll try it.

**Comment by maximbaz** [comment]
I have created #256, but then found out that the change comes from this PR 🙂 
Is there any chance to have icons on Windows without breaking the icons on Linux?

**Comment by SkaceKamen** [comment]
Damn, I didn't test it on linux, will take a look.

**Comment by SkaceKamen** [comment]
The electron method works fine, shall I create merge request with this method?

**Comment by Stanzilla** [comment]
@SkaceKamen yes! :)

**Comment by KELiON** [comment]
@SkaceKamen awesome! I'll wait for your PR!

**Comment by KELiON** [comment]
Hey, @SkaceKamen will you do PR with changed API for windows icons fetch? If you don't have time – can you just push somewhere your test branch?

**Comment by SkaceKamen** [comment]
Sorry, I just yesterday returned from vacation, I'll create PR right now.

**Comment by KELiON** [comment]
No problem, I hope you had a great time! :)

**Pull request merged**